mne/preprocessing/maxwell.py

```diff
@@ -0,0 +1,530 @@
+# Authors: Mark Wronkiewicz <wronk.mark@gmail.com>
+#          Jussi Nurminen <jnu@iki.fi>
+
+
+# License: BSD (3-clause)
+
+from __future__ import division
+from os import path as op
+import warnings
+import numpy as np
+from scipy.linalg import pinv
+from scipy.misc import factorial
+
+from .. import pick_types, pick_info
+from ..io.constants import FIFF
+from ..forward._compute_forward import _concatenate_coils
+from ..forward._make_forward import _read_coil_defs, _create_coils
+
+
+def maxwell_filter(raw, origin=(0, 0, 40), int_order=8, ext_order=3):
+    """Apply Maxwell filter to data using spherical harmonics.
+
+    Parameters
+    ----------
+    raw : instance of mne.io.Raw
+        Data to be filtered
+    origin : array-like, shape (3,)
+        Origin of internal and external multipolar moment space in head coords
+        and in millimeters
+    int_order : int
+        Order of internal component of spherical expansion
+    ext_order : int
+        Order of external component of spherical expansion
+
+    Returns
+    -------
+    raw_sss : instance of mne.io.Raw
+        The raw data with Maxwell filtering applied
+
+    Notes
+    -----
+    .. versionadded:: 0.10
+
+    Equation numbers refer to Taulu and Kajola, 2005.
+
+    There are an absurd number of different possible notations for spherical
+    coordinates, which confounds the notation for spherical harmonics.  Here,
+    we purposefully stay away from shorthand notation in both and use explicit
+    terms (like 'azimuth' and 'polar') to avoid confusion.
+
+    This code was adapted and relicensed (with BSD form) with permission from
+    Jussi Nurminen.
+
+    References
+    ----------
+    .. [1] Taulu and Kajola, 2005. "Presentation of electromagnetic
+           multichannel data: The signal space separation method".
+           http://lib.tkk.fi/Diss/2008/isbn9789512295654/article2.pdf
+    """
+
+    # TODO: Exclude 'bads' in multipolar moment calc, add back during
+    # reconstruction
+    if len(raw.info['bads']) > 0:
+        raise RuntimeError('Maxwell filter does not yet handle bad channels.')
+
+    # TODO: Improve logging process to better match Elekta's version
+    # Read coil definitions from file
+    all_coils, meg_info = _make_coils(raw.info, accurate=True,
+                                      elekta_defs=True)
+
+    # Create coil list and pick MEG channels
+    picks = [raw.info['ch_names'].index(coil['chname'])
+             for coil in all_coils]
+    coils = [all_coils[ci] for ci in picks]
+    raw.pick_channels(coil['chname'] for coil in coils)
+
+    data, times = raw[picks, :]
```

Comment:
We need to ensure that the data are not projected before grabbing the raw data. Maybe check `raw.proj` and throw and error if projections are being done? I thought they would be by default. But I also think I brought this up before and was surprised by the result. In any case, please check and then add a comment here about this issue.